server/sonar-server/src/main/java/org/sonar/server/measure/ws/SearchAction.java

```diff
@@ -0,0 +1,214 @@
+
+  private final DbClient dbClient;
+
+  public SearchAction(DbClient dbClient) {
+    this.dbClient = dbClient;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction("search")
+      .setInternal(true)
+      .setDescription("Search for component measures ordered by component names.<br>" +
+        "At most %d components can be provided.<br>" +
+        "Either '%s' or '%s' must be provided, not both.<br>" +
+        "Requires one of the following permissions:" +
+        "<ul>" +
+        " <li>'Administer System'</li>" +
+        " <li>'Administer' rights on the provided components</li>" +
+        " <li>'Browse' on the provided components</li>" +
+        "</ul>",
+        MAX_NB_COMPONENTS, PARAM_COMPONENT_IDS, PARAM_COMPONENT_KEYS)
+      .setSince("6.1")
+      .setResponseExample(getClass().getResource("search-example.json"))
+      .setHandler(this);
+
+    createMetricKeysParameter(action);
+
+    action.createParam(PARAM_COMPONENT_IDS)
+      .setDescription("Comma-separated list of component ids")
+      .setExampleValue(String.join(",", UUID_EXAMPLE_01, UUID_EXAMPLE_02, UUID_EXAMPLE_03));
+
+    action.createParam(PARAM_COMPONENT_KEYS)
+      .setDescription("Comma-separated list of component keys")
+      .setExampleValue(String.join(",", KEY_PROJECT_EXAMPLE_001, KEY_FILE_EXAMPLE_001, KEY_PROJECT_EXAMPLE_002, KEY_FILE_EXAMPLE_002));
+
+  }
+
+  @Override
+  public void handle(Request httpRequest, Response httpResponse) throws Exception {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      SearchWsResponse response = new ResponseBuilder(httpRequest, dbSession).build();
+      writeProtobuf(response, httpRequest, httpResponse);
+    }
+
+  }
+
+  private class ResponseBuilder {
+    private final DbSession dbSession;
+    private final Request httpRequest;
+    private SearchRequest request;
+    private List<ComponentDto> components;
+    private List<MetricDto> metrics;
+    private List<MeasureDto> measures;
+
+    ResponseBuilder(Request httpRequest, DbSession dbSession) {
+      this.dbSession = dbSession;
+      this.httpRequest = httpRequest;
+    }
+
+    SearchWsResponse build() {
+      this.request = setRequest();
+      this.components = searchComponents();
+      this.metrics = searchMetrics();
+      this.measures = searchMeasures();
+
+      return buildResponse();
+    }
+
+    private SearchRequest setRequest() {
+      request = SearchRequest.builder()
+        .setMetricKeys(httpRequest.mandatoryParamAsStrings(PARAM_METRIC_KEYS))
+        .setComponentIds(httpRequest.paramAsStrings(PARAM_COMPONENT_IDS))
+        .setComponentKeys(httpRequest.paramAsStrings(PARAM_COMPONENT_KEYS))
+        .build();
+
+      this.components = searchComponents();
+      this.metrics = searchMetrics();
+
+      return request;
+    }
+
+    private List<MetricDto> searchMetrics() {
+      requireNonNull(request);
```

Comment:
in fact I have seen the light and we should separate 2 steps: searching data and building the response. I'm talking about quite complex domain